src/vs/editor/common/modes.ts

```diff
@@ -821,8 +821,22 @@ export interface IResourceEdit {
 	range: IRange;
 	newText: string;
 }
+
+export interface IResourceRename {
+	readonly from: URI;
+	readonly to: URI;
+}
+
+export interface IResourceCreate {
+	readonly uri: URI;
+	readonly contents: string;
```

Comment:
Is `contents` field really necessary? WorkspaceEdit also has edits that can be applied to 0,0 location.

Reply:
Yeah, tempting... We could then just have one type, a rename is `from: Uri, to: Uri`, a create would be `from: null, to: Uri`, and a delete would be `from: Uri, to: null`... Similar to how we map inserted, deleted, and changed text